File: Learn_DataStructure/Graph/DirectedDFS.java
package Graph;

/**
 * Created by sunkai on 2017/9/5
 */
public class DirectedDFS {
    private boolean[] marked;
    public DirectedDFS(Diagraph G,int s){
        marked=new boolean[G.V()];
        dfs(G,s);
    }
    public DirectedDFS(Diagraph G,Iterable<Integer> sources){
        marked=new boolean[G.V()];
        for(int s:sources){
            if(!marked[s]){
                dfs(G,s);
            }
        }
    }
    private void dfs(Diagraph G,int v){
        marked[v]=true;
        for(int w:G.adj(v)){
            if(!marked[w]){
                dfs(G,w);
            }
        }
    }
    public boolean marked(int v){
        return marked[v];
    }
    public static void main(String args[]){

    }
}

File: Algorithm_Exercises/Algorithm (4th edition)/Algorithm/Chapter4/DiGraph/DirectedDFS.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Chapter4.DiGraph
{
    public class DirectedDFS
    {
        private bool[] marked;
        //在G中找到s可达的所有顶点
        public DirectedDFS(IDiGraph G, int s)
        {
            marked = new bool[G.V];
            Dfs(G, s);
        }

        //在G中找到从source中的所有顶点可达的所有顶点
        public DirectedDFS(IDiGraph G, IEnumerable<int> sources)
        {
            marked = new bool[G.V];
            foreach (int s in sources)
                if (!marked[s])
                    Dfs(G, s);
        }

        private void Dfs(IDiGraph g, int s)
        {
            marked[s] = true;
            foreach (int w in g.Adj(s))
                if (!marked[w])
                    Dfs(g, w);
        }

        public bool Marked(int v) => marked[v];

    }
}
